sdk/core/azure-core/src/main/java9/com/azure/core/implementation/ReflectionUtils.java

```diff
@@ -0,0 +1,65 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.core.implementation;
+
+import java.lang.invoke.MethodHandles;
+
+/**
+ * Utility methods that aid in performing reflective operations.
+ */
+public final class ReflectionUtils {
+    // This lookup is specific to the com.azure.core module, specifically this class.
+    private static final MethodHandles.Lookup LOOKUP = MethodHandles.lookup();
+
+    // Convenience pointer to the com.azure.core module.
+    private static final Module CORE_MODULE = ReflectionUtils.class.getModule();
+
+    /**
+     * Gets the {@link MethodHandles.Lookup} to use when performing reflective operations.
+     * <p>
+     * If Java 8 is being used this will always return {@link MethodHandles.Lookup#publicLookup()} as Java 8 doesn't
+     * have module boundaries that will prevent reflective access to the {@code targetClass}.
+     * <p>
+     * If Java 9 or above is being used this will return a {@link MethodHandles.Lookup} based on whether the module
+     * containing the {@code targetClass} exports the package containing the class. Otherwise, the {@link
+     * MethodHandles.Lookup} associated to {@code com.azure.core} will attempt to read the module containing {@code
```

Comment:
Does the JavaDoc have to include comments for both Java 8 and 9? Should this only have comments about Java 9 as this class is specific to this version? I am not sure what the IDE experience will be when you look at the docs for a specific method. Ideally, it should resolve the comments on the class that applies to the Java version used in the IDE.

Reply:
Luckily this is an implementation method, so the Javadocs here aren't a large concern, but I'll get them cleaned up.

For the question on how will the docs be presented when using the JAR as a dependency, I'll take a look into this.

Reply:
Unrelatedly, I get anxious about not having 100% overlap or mistakenly editing one file without realising that all files need to be kept in sync. We should consider developing a simple maven build tool that finds all such MR-jar 'extension' methods and compares the API signatures between them.

Reply:
> Unrelatedly, I get anxious about not having 100% overlap or mistakenly editing one file without realising that all files need to be kept in sync. We should consider developing a simple maven build tool that finds all such MR-jar 'extension' methods and compares the API signatures between them.

If testing is adequate, not having a complete API convergence between supported versions will result in a `MethodNotFound` exception.

Reply:
I was thinking about this yesterday. We could also consider introducing an interface that is implemented by the MR classes, and then have the Java compiler enforce this.

Reply:
> I was thinking about this yesterday. We could also consider introducing an interface that is implemented by the MR classes, and then have the Java compiler enforce this.

At least when it comes to publicly exposed multi-release APIs I agree that an interface will be the best option available, do we want to have the same pattern for internal APIs as well?

Reply:
Right, interface everywhere might help to keep us honest.